### DISCUSSION TITLE
Kraft v2 "Roadmap"

### DISCUSSION THREAD

**Discussion by dragotin** [dragotin/kraft#197] [Roadmap]
After having Kraft 1.0 released, this should outline a plan for Kraft 2.0 and add more transparency to where Kraft is moving. 

Please note: This is always subject to change! Also, the following list is probably too big and might be split to more major versions.

These are the big building blocks that are on the list:

## Technical Debt

Here we talk about technical topics that need to be done just to keep up with the fast moving ecosystem.

- Qt 6 port - Kraft 2.0 will (probably) build only against Qt6.
- drop ctemplate - not well maintained, move to grantlee everywhere.
- drop ReportLab - not well maintained, move to weasyprint complately.

## Projects

A simple yet useful Project management that adds more organisation around connected documents (think offer->partial invoice->invoice) and for example payment information.

## Private Clouds

Kraft should optionally integrate with *private* clouds such as ownCloud and Nextcloud. In fact, it does already today with the addressbook integration via Akonadi. But there could be much more, for example:

- Multi user: Multiple users create documents in parallel from different places. All is synchronized over "the cloud".
- Simple mobile access to documents and catalogs
- Kraft could create a "project space" with secret links for the customer where documents, photos etc. can be exchanged.

This has some technical implications such as: For example, the database should become a local storage only (Sqlite only) and there should not be a need to synchronize the DB.
 
## Address Book Flexibility

The current Akonadi based addressbook is great and will continue to be the main integration. The problem with it is that it does not work with the AppImage. There needs to be a solution for this, most likely file based.

## Features

There are many [enhancement requests](https://github.com/dragotin/kraft/issues?q=is%3Aopen+is%3Aissue+label%3Aenhancement) in the issue tracker.

Give your vote setting a :+1:  or :heart:  emoji to the enhancements you would consider most valueable.

**Comment by dragotin** [comment]
Kraft v2 is progressing in the branch _newshared_: https://github.com/dragotin/kraft/tree/newshared

A few more features that already landed in the branch:

## Document Livecycle

In Kraft v2 there will be a new livecycle for documents. 

These are the states defined (today):

* **Undefined**:  State not defined at all. Internal program state
* **New**: New document, not yet saved to disk
* **Draft**: New document, saved on disk, actively worked on but not yet sent to customer. Does not have an official document number yet, and can be deleted as a result
* **Final**: Saved and with an official doc number, equivalent to sent out. Should/can not be deleted
* **Retracted**: Sent to customer, but retracted, follows state Final
* **Converted**:  Converted from a previous Kraft version that did not have proper states
* **Invalid**: Invalidated. Never sent out

Downside is that users have to do a finalizing step to make the document final. However, that comes with many benefits, for example simpler deletion of draft docs and reminders for unpaid invoices.

## XML Document Based

To be able to sync documents to private clouds, Kraft v2 is not longer storing documents in the database, but as individual XML documents in the file system. A file system structure and a file index for quick lookup of documents are already there.

Also the number cycles are now stored in an XML file.

## Prompt PDF Creation

Kraft v2 generates the PDF not at the moment where user wants to open the document, but as soon as a change to the document is saved to disk. That gives a much better user experience because the PDF is opened immediately.

## Visual Changes

Use a consistent icon set (Tabler icons) also on the document overview pane.
Added a welcome text after startup.
![image](https://github.com/dragotin/kraft/assets/1070214/26d0a1f9-fe81-4b94-b59e-934f9aea7c46)

**Comment by dragotin** [comment]
:rocket:  The Kraft version 2 feature branch called _newshared_ was merged into master today. :tada: 

The next version that will be released from master is a Kraft version 2.x. It is not yet clear when that will be. For ongoing work on Kraft version 1.x the branch [`rel1.2`](https://github.com/dragotin/kraft/tree/rel1.2) has to be used.

**Comment by dragotin** [comment]
Address book in AppImage - code has landed today in branch [portqt6_addrdir](https://github.com/dragotin/kraft/tree/portqt6_addrdir).

It is an AddressProvider that reads vcf-files from a local directory and uses them as address book in Kraft. The local directory can either be filled by the "Personal Addressbook" in Kaddressbook, or also with a tool like [vdirsyncer](https://github.com/pimutils/vdirsyncer) and thus use Google addressbook if needed.

This is a big step forward.

**Comment by szotsaki** [comment]
May I ask you to update the [openSUSE package](https://build.opensuse.org/package/show/openSUSE%3AFactory/kraft) to version 2.0? It seems it still offers 1.2.2.

**Comment by dragotin** [comment]
It is in KDE:Extra now, and on the way to Factory aka tumbleweed

**Comment by dragotin** [comment]
Closing this discussion as Kraft 2 is released now. See [new discussion topic](https://github.com/dragotin/kraft/discussions/278) to discuss the new release.
